src/AutoRest.CSharp.V3/Output/Builders/SerializationBuilder.cs

```diff
@@ -150,24 +150,70 @@ private static XmlObjectSerialization BuildXmlObjectSerialization(ObjectSchema o
+        {
+            foreach (Property property in propertyBag.Properties)
+            {
+                yield return new JsonPropertySerialization(
+                    property.SerializedName,
+                    property.CSharpName(),
+                    BuildSerialization(property.Schema, property.IsNullable()),
+                    false);
+            }
+
+            foreach ((string name, PropertyBag innerBag) in propertyBag.Bag)
+            {
+                JsonPropertySerialization[] serializationProperties = GetPropertySerializationsFromBag(innerBag).ToArray();
+                JsonObjectSerialization objectSerialization = new JsonObjectSerialization(new SchemaTypeReference(null, false), serializationProperties, null);
+                yield return new JsonPropertySerialization(name, name, objectSerialization, true);
+            }
+        }
+
         private static JsonObjectSerialization BuildJsonObjectSerialization(ObjectSchema objectSchema, TypeReference schemaTypeReference, bool isNullable)
         {
             List<JsonPropertySerialization> serializationProperties = new List<JsonPropertySerialization>();
             foreach (var schema in EnumerateHierarchy(objectSchema))
             {
-                foreach (Property property in schema.Properties!)
+                PropertyBag propertyBag = new PropertyBag();
+                propertyBag.Properties.AddRange(schema.Properties ?? Array.Empty<Property>());
+                PopulatePropertyBag(propertyBag, 0);
+                serializationProperties.AddRange(GetPropertySerializationsFromBag(propertyBag));
+            }
+
+            return new JsonObjectSerialization(schemaTypeReference, serializationProperties.ToArray(), CreateAdditionalProperties(objectSchema));
+        }
+
+        private class PropertyBag
+        {
+            public Dictionary<string, PropertyBag> Bag { get; } = new Dictionary<string, PropertyBag>();
+            public List<Property> Properties { get; } = new List<Property>();
+        }
+
+        private static void PopulatePropertyBag(PropertyBag propertyBag, int depthIndex)
+        {
+            foreach (Property property in propertyBag.Properties.ToArray())
+            {
+                int flattenNamesCount = property.FlattenedNames?.Count ?? 0;
+                if (flattenNamesCount == 0 || depthIndex >= flattenNamesCount - 1)
                 {
-                    JsonPropertySerialization propertySerialization = new JsonPropertySerialization(
-                        property.SerializedName,
-                        property.CSharpName(),
-                        BuildSerialization(property.Schema, property.IsNullable())
-                    );
+                    continue;
+                }
 
-                    serializationProperties.Add(propertySerialization);
+                string name = property!.FlattenedNames.ElementAt(depthIndex);
```

Comment:
Why do you need ! here?

Reply:
Because we do `property.FlattenedNames?.Count ?? 0` before this, it can't guarantee that FlattenNames is not null. However, our condition itself (logically) means that we have FlattenNames to work with.